packages/tasit-action/src/TasitAct.js

```diff
@@ -0,0 +1,144 @@
+import "ethers/dist/shims.js";
+// Note: ethers SHOULD be imported from their main object
+// shims aren't injected with package import
+import { ethers } from "ethers";
+
+class Utils {
+  static isAddress = address => {
+    return typeof address === "string" && address.match(/^0x[0-9A-Fa-f]{40}$/);
+  };
+
+  static isABI = abi => {
+    return abi && Array.isArray(abi);
+  };
+
+  // https://github.com/ethers-io/ethers.js/blob/db383a3121bb8cf5c80c5488e853101d8c1df353/src.ts/utils/properties.ts#L20
+  static isEthersJsSigner = signer => {
+    return signer && signer._ethersType === "Signer";
+  };
+}
+
+class Subscription {
+  #txPromise;
+  #provider;
+
+  constructor(txPromise, provider) {
+    this.#txPromise = txPromise;
+    this.#provider = provider;
+  }
+
+  on = async (trigger, callback) => {
+    let blockHeight = 0;
+
+    if (trigger === "confirmation") blockHeight = 1;
+    else {
+      throw new Error(`Invalid subscription trigger, use: ["confirmation"]`);
+    }
+
+    if (!callback || typeof callback !== "function") {
+      throw new Error(`Cannot subscribe without a function`);
+    }
+
+    const tx = await this.#txPromise;
+
+    this.#provider.on(tx.hash, async receipt => {
+      const message = {
+        data: {
+          confirmations: receipt.confirmations,
+        },
+      };
+
+      try {
+        await callback(message);
+      } catch (error) {
+        throw new Error(`Callback function with error: ${error.message}`);
```

Comment:
I wonder if we want to emit an error event on the subscription?

Reply:
Something like this? [web3.js 1.0 API](https://web3js.readthedocs.io/en/1.0/web3-eth-subscribe.html#id11)
```javascript
var subscription = web3.eth.subscribe('newBlockHeaders', function(error, result){
    if (!error) {
        console.log(result);

        return;
    }

    console.error(error);
})
.on("data", function(blockHeader){
    console.log(blockHeader);
})
.on("error", console.error);

// unsubscribes the subscription
subscription.unsubscribe(function(error, success){
    if (success) {
        console.log('Successfully unsubscribed!');
    }
});
```

Reply:
Yeah. I personally would add the error on a separate line rather than chaining, but that's a matter of personal preference I guess.

The main difference is that I think the errors here are things that the subscription itself deems errors rather than errors in the function that handles incoming data.

But we could have errors in the function that handles incoming data emit error events that are then handled by the other handler.

Also I would have the subscribe and unsubscribe functions use await style rather than callbacks.

Reply:
I think that was solved after the introduction of "error" event. Is it true?

Reply:
A) Chaining is solved with that, yes.

B) And using await still with subscribe and unsubscribe is no longer relevant if the subscription is sync.

C) There's one other topic here, that I extracted - this issue I created for making sure we pass through errors that don't come from listeners: https://github.com/tasitlabs/TasitSDK/issues/124

Reply:
Can we solve this?

Reply:
Yep! Resolving.